### PR TITLE
Source maps

### PR DESCRIPTION
**Pull request by rqu**
- [x] source line mapping
- [x] breakpoints
- [x] ~~method name mapping (if possible)~~
- [x] variable name mapping (if possible)

### CODE REVIEW

**Comment by rqu** [comment]
I omitted stepping from the list. It does not work very well and it probably won't. When attempting to step-over a statement, it sometimes drops to the compiled source because of a "missing" mapping for statements like `if(gt <= 42) ...`. I assume `gt` here stands for `goto`. Mapping these to the following instructions makes breakpoints unreliable and does not help much with the stepping (each statement is visited multiple times).

File: rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java [hunk 1]
```diff
@@ -43,6 +43,7 @@ private Main() {}
     public static void main(String... args) throws IOException, URISyntaxException {
         final String obfuscate = "--obfuscatelevel";
         final String extension = "--createextension";
+        final String sourcemap = "--sourcemap";
```
Comment (jtulach) [inline rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java:46]: I have forgotten there is a `main` method in the JAR! I assume you are using this `main` for testing, right? What are the CLI commands you use?

It'd be good to integrate (or describe how such integration should look like) with Maven and test infrastructure. 

There are two versions when one generates `-bck2brwsr.jar` artifacts with pre-compiled Bck2Brwsr `.js` files - production (usually obfuscated) and debugging one. It'd be great if the source maps (and sources?) could be attached to the debug one.

Then there is the `VMTest` functionality - it should generate source maps by default and make them available when the test is running in the browser.

**Review by jtulach — COMMENTED** [review]
Thanks a lot for the contribution Lukáši. I am eager to polish it and make it available in next version of Bck2Brwsr.

**Comment by rqu** [comment]
Method names are honored by Node.js only, which is not exactly our target group. Firefox and Chromium ignore them in stack traces and debugger stack view. Hence I omit them from the source map.

Note that it is possible to name a function even in a function expression for example like this: `m = function MyClass_myMethod(...) { ... }`. This introduces a name only in the inner function's own scope. We would have to take care to (1) never emit a keyword and (2) never hide an outside identifier we want to use. Do we want to do this @jtulach?

I have also tried to assign the `m.name` property, but the assignment has no effect (not even changing the property).

**Comment by rqu** [comment]
Variable names won't work in Firefox. Firefox DevTools has two restrictions: it cannot parse Java files, and it ignores variable name mappings in source files it cannot parse (i.e. both unsupported and supported but invalid).

OTOH, Chromium DevTools can parse Java (or at least highlight it) and even if it could not (invalid or too new Java, or unknown JVM language and a different file extension), it still maps the variable names as per the source map.

File: rt/vm/src/main/java/org/apidesign/vm4brwsr/Bck2Brwsr.java [hunk 2]
```diff
@@ -344,12 +344,12 @@ public void generate(Appendable out) throws IOException {
         VM.compile(out, null, this);
     }
 
-    void generate(Appendable out, SourceMapBuilder smb) throws IOException {
+    void generate(Appendable out, SourceMapGenerator srcmap) throws IOException {
```
Comment (rqu) [inline rt/vm/src/main/java/org/apidesign/vm4brwsr/Bck2Brwsr.java:347]: Didn't like what `smb` sounded like. :police_officer:

File: rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java (hunk 1 above)
Comment (jtulach) [reply rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java:46]: Heuréka. I've got it working with:
```
bck2brwsr$ export R=`pwd`
bck2brwsr/javaquery/demo-calculator/target/demo.static.calculator-1.0-SNAPSHOT-bck2brwsr/public_html/app$ ln -s ../../../../src/main/java/org/ .
bck2brwsr/javaquery/demo-calculator$ java -ea -cp $R/rt/emul/compact/target/emul-1.0-SNAPSHOT.jar:$R/rt/emul/mini/target/emul.mini-1.0-SNAPSHOT.jar:$R/rt/vm/target/vm4brwsr-1.0-SNAPSHOT.jar:target/demo.static.calculator-1.0-SNAPSHOT.jar org.apidesign.vm4brwsr.Main --createextension --obfuscatelevel NONE --sourcemap target/demo.static.calculator-1.0-SNAPSHOT-bck2brwsr/public_html/app/calculator.js org/apidesign/bck2brwsr/demo/calc/staticcompilation/Preinit 
bck2brwsr/javaquery/demo-calculator$ mvn bck2brwsr:show
```
I am able to regenerate the `calculator.js` and put breakpoint into `Preinit.sayHello` Java source code. The chrome debugger stops there and shows value of `msg` parameter. Great work!

Comment (jtulach) [reply rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java:46]: Please accept this https://github.com/jtulach/bck2brwsr/commit/d027e0089f309e265d925772a7f27934fde15e1c into your PR. It is adding `Bck2Brwsr.generate(Appendable, String, Appendable)`. Next I plan to use it from Maven plugin.

Comment (jtulach) [reply rt/vm/src/main/java/org/apidesign/vm4brwsr/Main.java:46]: With https://github.com/jtulach/bck2brwsr/commit/1ddac1ef970750f48b7ceb2e96e18c563c132142 I can build `bck2brwsr` and then:
```
bck2brwsr$ mvn -f javaquery/demo-calculator/ clean package bck2brwsr:show \
    -Dbck2brwsr.obfuscationlevel=NONE -DskipTests
```
the calculator application is shown and debugging works like a charm!

image

**Pull request merged**